protocols/gossipsub/src/behaviour.rs

```diff
@@ -3071,15 +2888,7 @@ where
                 if !subscriptions.is_empty() {
                     // send our subscriptions to the peer
                     if self
-                        .send_message(
-                            peer_id,
-                            Rpc {
-                                messages: Vec::new(),
-                                subscriptions,
-                                control_msgs: Vec::new(),
-                            }
-                            .into_protobuf(),
-                        )
+                        .send_message(peer_id, RpcOut::Subscriptions(subscriptions))
```

Comment:
Same as above, instead of collecting in a temporary `Vec`, lets call `self.send_message` for each subscription that is pending.

Reply:
Another refactoring we can do here is to be more defensive with our `if`s:

```rust
if other_established > 0 {
	return; // Not our first connection to this peer, hence nothing to do.
}

if self.blacklisted_peers.contains(&peer_id) {
	tracing::debug!(peer=%peer_id, "Ignoring connection from blacklisted peer");
	return;
}
```

YMMV but I much prefer this style because it reduces nesting and allows my brain to think through these conditions individually instead of all at once.

Reply:
Ideally, we would move the sending of pending subscriptions to the `handle_established_{in,out}bound_connection` callback and directly give them to the handler. That way, it can immediately start processing these messages instead of injecting them here in the callback.

This wasn't as easily possible when this code was written because the previous `new_handler` function had an indirection until it created the `ConnectionHandler`.

This is definitely a refactoring that we can do at a later point though.

Reply:
yeah, agree and updated, but the second case you mention I also noticed and I agree that we should return earlier but that is different behaviour than what we currently have ptal Thomas, did it on an isolated commit:  f2b1f420368e197bc6052dbfb4d47c8f014616f2

Reply:
> but that is different behaviour than what we currently have

You can restore the old behaviour but just moving everything that needs to always happen before the early return :)
That is nice thing about this style, it makes it explicit that something happens independently of a certain condition instead of having to parse in your head, where certain code branches add and which other conditions are essentially implicit.